managed/services/telemetry/datasource_grafana_sqlitedb.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package telemetry provides telemetry functionality.
+package telemetry
+
+import (
+	"context"
+	"database/sql"
+	"io"
+	"os"
+
+	// Driver for for grafana sqlite database.
+	_ "github.com/mattn/go-sqlite3"
+	pmmv1 "github.com/percona-platform/saas/gen/telemetry/events/pmm"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type dsGrafanaSelect struct {
+	log      *logrus.Entry
+	config   DSGrafanaSqliteDB
+	db       *sql.DB
+	tempFile string
+}
+
+// check interfaces.
+var (
+	_ DataSource = (*dsGrafanaSelect)(nil)
+)
+
+// Enabled flag that determines if data source is enabled.
+func (d *dsGrafanaSelect) Enabled() bool {
+	return d.config.Enabled
+}
+
+// NewDataSourceGrafanaSqliteDB makes new data source for grafana sqlite database metrics.
+func NewDataSourceGrafanaSqliteDB(config DSGrafanaSqliteDB, l *logrus.Entry) DataSource {
+	return &dsGrafanaSelect{
+		log:      l,
+		config:   config,
+		db:       nil,
+		tempFile: "",
+	}
+}
+
+func (d *dsGrafanaSelect) PreFetch(ctx context.Context) error {
+	// validate source file db
+	sourceFileStat, err := os.Stat(d.config.DBFile)
+	if err != nil {
+		return err
+	}
+
+	if !sourceFileStat.Mode().IsRegular() {
+		return errors.Wrapf(err, "%s is not a regular file", d.config.DBFile)
+	}
+
+	source, err := os.Open(d.config.DBFile)
+	if err != nil {
+		d.log.Error(err)
+		return err
+	}
+
+	tempFile, err := os.CreateTemp(os.TempDir(), "grafana")
+	if err != nil {
+		d.log.Error(err)
+		return err
+	}
+	defer func() {
+		if err := source.Close(); err != nil {
+			d.log.Errorf("Error closing file. %s", err)
+		}
+	}()
+
+	nBytes, err := io.Copy(tempFile, source)
+	if err != nil || nBytes == 0 {
```

Comment:
I understand that `nBytes == 0` should not happen, but what if source file is empty?

I think we should state "nBytes == 0 && sourcefileLen > 0"

Reply:
Fixed for both cases.